src/analyzer/plugins/analyzerkeyfinder.cpp

```diff
@@ -0,0 +1,115 @@
+
+#include "analyzer/constants.h"
+#include "util/assert.h"
+#include "util/math.h"
+
+using mixxx::track::io::key::ChromaticKey;
+using mixxx::track::io::key::ChromaticKey_IsValid;
+
+namespace {
+ChromaticKey chromaticKeyFromKeyFinderKeyT(KeyFinder::key_t key) {
+    switch (key) {
+    case (KeyFinder::A_MAJOR):
+        return ChromaticKey::A_MAJOR;
+    case (KeyFinder::A_MINOR):
+        return ChromaticKey::A_MINOR;
+    case (KeyFinder::B_FLAT_MAJOR):
+        return ChromaticKey::B_FLAT_MAJOR;
+    case (KeyFinder::B_FLAT_MINOR):
+        return ChromaticKey::B_FLAT_MINOR;
+    case (KeyFinder::B_MAJOR):
+        return ChromaticKey::B_MAJOR;
+    case (KeyFinder::B_MINOR):
+        return ChromaticKey::B_MINOR;
+    case (KeyFinder::C_MAJOR):
+        return ChromaticKey::C_MAJOR;
+    case (KeyFinder::C_MINOR):
+        return ChromaticKey::C_MINOR;
+    case (KeyFinder::D_FLAT_MAJOR):
+        return ChromaticKey::D_FLAT_MAJOR;
+    case (KeyFinder::D_FLAT_MINOR):
+        return ChromaticKey::C_SHARP_MINOR;
+    case (KeyFinder::D_MAJOR):
+        return ChromaticKey::D_MAJOR;
+    case (KeyFinder::D_MINOR):
+        return ChromaticKey::D_MINOR;
+    case (KeyFinder::E_FLAT_MAJOR):
+        return ChromaticKey::E_FLAT_MAJOR;
+    case (KeyFinder::E_FLAT_MINOR):
+        return ChromaticKey::E_FLAT_MINOR;
+    case (KeyFinder::E_MINOR):
+        return ChromaticKey::E_MINOR;
+    case (KeyFinder::F_MAJOR):
+        return ChromaticKey::F_MAJOR;
+    case (KeyFinder::F_MINOR):
+        return ChromaticKey::F_MINOR;
+    case (KeyFinder::G_FLAT_MAJOR):
+        return ChromaticKey::F_SHARP_MAJOR;
+    case (KeyFinder::G_FLAT_MINOR):
+        return ChromaticKey::F_SHARP_MINOR;
+    case (KeyFinder::G_MAJOR):
+        return ChromaticKey::G_MAJOR;
+    case (KeyFinder::G_MINOR):
+        return ChromaticKey::G_MINOR;
+    case (KeyFinder::A_FLAT_MAJOR):
+        return ChromaticKey::A_FLAT_MAJOR;
+    case (KeyFinder::A_FLAT_MINOR):
+        return ChromaticKey::G_SHARP_MINOR;
+    case (KeyFinder::SILENCE):
+    default:
+        return ChromaticKey::INVALID;
+    }
+}
+} // anonymous namespace
+
+namespace mixxx {
+
+AnalyzerKeyFinder::AnalyzerKeyFinder()
+        : m_keyFinder(),
+          m_workspace(),
+          m_audioData(),
+          m_currentFrame(0),
+          m_previousKey(mixxx::track::io::key::INVALID),
+          m_resultKeys() {
+}
+
+bool AnalyzerKeyFinder::initialize(int samplerate) {
+    m_audioData.setFrameRate(samplerate);
+    m_audioData.setChannels(kAnalysisChannels);
+    return true;
+}
+
+bool AnalyzerKeyFinder::processSamples(const CSAMPLE* pIn, const int iLen) {
+    DEBUG_ASSERT(iLen % kAnalysisChannels == 0);
+    if (m_audioData.getSampleCount() == 0) {
+        m_audioData.addToSampleCount(iLen);
+    }
+
+    const SINT numInputFrames = iLen / kAnalysisChannels;
+    m_currentFrame += numInputFrames;
+
+    for (SINT frame = 0; frame < numInputFrames; frame++) {
```

Comment:
This is what I had in mind for destructuring the interlaced sample buffer:
```
for (SINT sample = 0; sample < iLen; ++sample) {
    m_audioData.setSampleByFrame(
            sample / kAnalysisChannels,
            sample % kAnalysisChannels,
            pIn[sample]);
}
```

Reply:
IMO that is harder to understand than the nested loop. If you really want a single loop we can change it to use KeyFinder::AudioData::setSample. But if we are going to remove the assumption of 2 channels from Mixxx, I think we will need these nested frame and channel loops most places we manipulate an audio buffer.

Reply:
Don't mind, was just an idea.